### ISSUE TITLE
yarn add does not work with private/proxy repository

### ISSUE THREAD

**Issue by bertramn** [yarnpkg/yarn#4932]
We are running a Sonatype Nexus Repository mirror and configuring yarn to use that mirror. Unfortunately yarn is unable to connect to this server and fails with `ETIMEDOUT`.

```sh
# yarn --version
1.3.2
```
**Do you want to request a *feature* or report a *bug*?**

bug

**What is the current behavior?**

Install nodejs ... we did install 8.9.1.

Point to an internal Nexus mirror

```sh
$ npm config set registry https://nexus.example.com/repository/external-npm-npmjs/ -g
```

npm config looks good

```sh
$ npm config list -l
; cli configs
long = true
metrics-registry = "https://nexus.example.com/repository/external-npm-npmjs/"
scope = ""
user-agent = "npm/5.5.1 node/v8.9.1 linux x64"

; globalconfig /opt/node-v8.9.1-linux-x64/etc/npmrc
registry = "https://nexus.example.com/repository/external-npm-npmjs/"
...
globalconfig = "/opt/node-v8.9.1-linux-x64/etc/npmrc"
globalignorefile = "/opt/node-v8.9.1-linux-x64/etc/npmignore"
prefix = "/opt/node-v8.9.1-linux-x64"
; registry = "https://registry.npmjs.org/" (overridden)
; user-agent = "npm/{npm-version} node/{node-version} {platform} {arch}" (overridden)
userconfig = "/root/.npmrc"
```

Install Yarn via npm, nicely working ...

```sh
$ npm install -g yarn
npm install -g yarn
/opt/node-v8.9.1-linux-x64/bin/yarn -> /opt/node-v8.9.1-linux-x64/lib/node_modules/yarn/bin/yarn.js
/opt/node-v8.9.1-linux-x64/bin/yarnpkg -> /opt/node-v8.9.1-linux-x64/lib/node_modules/yarn/bin/yarn.js
+ yarn@1.3.2
added 1 package in 2.686s
```

Configure Yarn

```sh
$ yarn config set registry 'https://nexus.example.com/repository/external-npm-npmjs/' -g
yarn config v1.3.2
success Set "registry" to "https://nexus.example.com/repository/external-npm-npmjs/".
Done in 0.11s.
```

Yarn config looks good

```
$ yarn config list

yarn config v1.3.2
info yarn config
{ 'version-tag-prefix': 'v',
  'version-git-tag': true,
  'version-git-sign': false,
  'version-git-message': 'v%s',
  'init-version': '1.0.0',
  'init-license': 'MIT',
  'save-prefix': '^',
  'ignore-scripts': false,
  'ignore-optional': false,
  registry: 'https://nexus.example.com/repository/external-npm-npmjs/',
  'strict-ssl': true,
  'user-agent': 'yarn/1.3.2 npm/? node/v8.9.1 linux x64',
  lastUpdateCheck: 1510618119102,
  test_val: true }
info npm config
{ registry: 'https://nexus.example.com/repository/external-npm-npmjs/' }
Done in 0.10s.
```

Lets try to add gulp via yarn

```sh
$ yarn add -g --verbose gulp
yarn add v1.3.2
verbose 0.672 Checking for configuration file "/root/.npmrc".
verbose 0.672 Checking for configuration file "/usr/local/share/.npmrc".
verbose 0.673 Checking for configuration file "/opt/node-v8.9.1-linux-x64/etc/npmrc".
verbose 0.675 Found configuration file "/opt/node-v8.9.1-linux-x64/etc/npmrc".
verbose 0.676 Checking for configuration file "/root/.npmrc".
verbose 0.677 Checking for configuration file "/root/.npmrc".
verbose 0.681 Checking for configuration file "/root/.yarnrc".
verbose 0.681 Checking for configuration file "/usr/local/share/.yarnrc".
verbose 0.682 Found configuration file "/usr/local/share/.yarnrc".
verbose 0.685 Checking for configuration file "/opt/node-v8.9.1-linux-x64/etc/yarnrc".
verbose 0.685 Checking for configuration file "/root/.yarnrc".
verbose 0.685 Checking for configuration file "/root/.yarnrc".
verbose 0.691 current time: 2017-11-14T00:33:51.044Z
info No lockfile found.
[1/4] Resolving packages...
verbose 0.844 Performing "GET" request to "https://nexus.example.com/repository/external-npm-npmjs/gulp".
verbose 30.872 Error: https://nexus.example.com/repository/external-npm-npmjs/gulp: ETIMEDOUT
    at Timeout._onTimeout (/opt/node-v8.9.1-linux-x64/lib/node_modules/yarn/lib/cli.js:123749:19)
    at ontimeout (timers.js:475:11)
    at tryOnTimeout (timers.js:310:5)
    at Timer.listOnTimeout (timers.js:270:5)
error An unexpected error occurred: "https://nexus.example.com/repository/external-npm-npmjs/gulp: ETIMEDOUT".
info If you think this is a bug, please open a bug report with the information provided in "/root/yarn-error.log".
info Visit https://yarnpkg.com/en/docs/cli/add for documentation about this command.
```

NOPE did not work... whats the error captain?

```sh
$ cat /root/yarn-error.log
Arguments:
  /opt/node-v8.9.1-linux-x64/bin/node /opt/node-v8.9.1-linux-x64/bin/yarn add -g --verbose gulp

PATH:
  /opt/node-v8.9.1-linux-x64/bin:/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin:/root/bin

Yarn version:
  1.3.2

Node version:
  8.9.1

Platform:
  linux x64

npm manifest:
  No manifest

yarn manifest:
  No manifest

Lockfile:
  No lockfile

Trace:
  Error: https://nexus.example.com/repository/external-npm-npmjs/gulp: ETIMEDOUT
      at Timeout._onTimeout (/opt/node-v8.9.1-linux-x64/lib/node_modules/yarn/lib/cli.js:123749:19)
      at ontimeout (timers.js:475:11)
      at tryOnTimeout (timers.js:310:5)
      at Timer.listOnTimeout (timers.js:270:5)
```

Meeh, but I just installed something from this very server. Also can confirm the box I tried this on has NO internet connectivity at all, just a firewall rule to allow it to connect to `nexus.example.com` on port 443.

We definitely have connectivity on this host.

```sh
$ curl https://nexus.example.com/repository/external-npm-npmjs/gulp
{"maintainers":[{"name":"contra","email":"yo@contra.io"},{"name":"phated","email":"blaine.bublitz@gmail.com"}],"dist-tags":{"latest":"3.9.1"},"
...
```

**What is the expected behavior?**

install gulp

**Please mention your node.js, yarn and operating system version.**

node 8.9.1
yarn 1.3.2
RHEL 7.2 x64 - run off the mill

**Comment by bertramn** [comment]
I did some more testing and just for giggles configured a proxy via `https_proxy` that curl and others would typically use. It turns out, once there is a usable proxy route, yarn shows that it is trying to connect to the mothership. So in the first example it probably does not even get as far as connecting to the actual registry.

```sh
$ export https_proxy='http://localhost:3128'

$ yarn add -g --verbose gulp

yarn add v1.3.2
verbose 0.35 Checking for configuration file "/root/.npmrc".
verbose 0.353 Checking for configuration file "/usr/local/share/.npmrc".
verbose 0.353 Checking for configuration file "/opt/node-v8.9.1-linux-x64/etc/npmrc".
verbose 0.354 Found configuration file "/opt/node-v8.9.1-linux-x64/etc/npmrc".
verbose 0.356 Checking for configuration file "/root/.npmrc".
verbose 0.357 Checking for configuration file "/root/.npmrc".
verbose 0.361 Checking for configuration file "/root/.yarnrc".
verbose 0.361 Checking for configuration file "/usr/local/share/.yarnrc".
verbose 0.361 Found configuration file "/usr/local/share/.yarnrc".
verbose 0.362 Checking for configuration file "/opt/node-v8.9.1-linux-x64/etc/yarnrc".
verbose 0.362 Checking for configuration file "/root/.yarnrc".
verbose 0.363 Checking for configuration file "/root/.yarnrc".
verbose 0.369 current time: 2017-11-15T04:42:08.332Z
info No lockfile found.
verbose 0.42 Performing "GET" request to "https://yarnpkg.com/latest-version". <<< !!!!!!!!!!!!!
[1/4] Resolving packages...
verbose 0.493 Performing "GET" request to "https://nexus.example.com/repository/external-npm-npmjs/gulp".
verbose 0.818 Request "https://yarnpkg.com/latest-version" finished with status code 200.
verbose 30.5 Error: https://nexus.example.com/repository/external-npm-npmjs/gulp: ETIMEDOUT
    at Timeout._onTimeout (/opt/node-v8.9.1-linux-x64/lib/node_modules/yarn/lib/cli.js:123749:19)
    at ontimeout (timers.js:475:11)
    at tryOnTimeout (timers.js:310:5)
    at Timer.listOnTimeout (timers.js:270:5)
error An unexpected error occurred: "https://nexus.example.com/repository/external-npm-npmjs/gulp: ETIMEDOUT".
info If you think this is a bug, please open a bug report with the information provided in "/root/yarn-error.log".
info Visit https://yarnpkg.com/en/docs/cli/add for documentation about this command.
```

**Comment by torifat** [comment]
Should be fixed in #4849 Feel free to reopen if you think it's a different issue.

**Comment by bertramn** [comment]
Ok thanks, confirming that downgrading to yarn 1.2.1 fixes this problem.
